server/src/main/java/org/elasticsearch/persistent/PersistentTasksService.java

```diff
@@ -192,17 +192,19 @@ public void waitForPersistentTaskCondition(
      * @param timeout a timeout for waiting
      * @param listener the callback listener
      */
-    public void waitForPersistentTaskCondition(
+    public <P extends PersistentTaskParams> void waitForPersistentTaskCondition(
         final ProjectId projectId,
         final String taskId,
-        final Predicate<PersistentTask<?>> predicate,
+        final Predicate<PersistentTask<P>> predicate,
         final @Nullable TimeValue timeout,
-        final WaitForPersistentTaskListener<?> listener
+        final WaitForPersistentTaskListener<P> listener
     ) {
         ClusterStateObserver.waitForState(clusterService, threadPool.getThreadContext(), new ClusterStateObserver.Listener() {
             @Override
             public void onNewClusterState(ClusterState state) {
-                listener.onResponse(PersistentTasksCustomMetadata.getTaskWithId(state.metadata().getProject(projectId), taskId));
+                final var project = state.metadata().projects().get(projectId);
+                final PersistentTask<P> task = project == null ? null : PersistentTasksCustomMetadata.getTaskWithId(project, taskId);
+                listener.onResponse(task);
```

Comment:
@ywangd I had to make some changes here to allow the extraction of the `task` variable. I had to explicitly type the `predicate` and `listener`, otherwise the compiler would complain. I think these changes make sense, but I'll let you have another look before I merge.

Reply:
If we don't have the separate `task` variable, we would not need the explicit type parameter changes. i.e. the following works

```java
listener.onResponse(project == null ? null : PersistentTasksCustomMetadata.getTaskWithId(project, taskId));
```

This is already the case with existing code, i.e. the current code does not compile if you have

```java
final var task = PersistentTasksCustomMetadata.getTaskWithId(state, taskId);
listener.onResponse(task);  // <--- compilation error
```

The wildcard capture of `listener.onReponse` does not work with other explicit type or another wildcard capture. It needs to infer it by itself.

So I think we can just inline the variable and no need for the type parameter changes.

Reply:
>  to allow the extraction of the task variable

If your goal is to add the separate `task` variable, the type parameter changes make sense. I don't feel the `task` variable is necessary. But I don't have strong opinion either.

Reply:
I did have a slight preference for the `task` variable. That combined with the fact that I think having the explicit type parameter makes sense here anyway, resulted in me doing the change this way. In multiple places, we already get (and cast) the state or params from the persistent task passed to the listener, so most usages already depend on the task being a specific instance. That said, I'm fine to revert the changes and inline the variable to reduce the changes of this commit a bit (i.e. keep the scope smaller).

Reply:
I see you opened #124180 with the project deletion checks. I'll just wait for you to merge that one and then merge this one. Then I don't have to add the project deletion checks in this PR.

Reply:
They are for different methods. I think there is no need to for any order between merging the two PRs.